doc/developer/design/20230728_left_join_stack_consolidation.md

```diff
@@ -0,0 +1,160 @@
+             TopK group_by=[#0] limit=1      +
+               Filter (#0) IS NOT NULL       +
+                 Get materialize.public.bar  + // 1, 1
+                                             +
+ Source materialize.public.bar               +
+   filter=((#0) IS NOT NULL)                 +
+ Source materialize.public.more1             +
+   filter=((#0) IS NOT NULL)                 +
+ Source materialize.public.more2             +
+   filter=((#0) IS NOT NULL)                 +
+```
+For each of the 3 LEFT JOINs, the plan has one `Union-Negate`: In `l1`, `l3`, and the `Let` body. The problem is that there is a dataflow path that has no consolidation between these Union-Negates: For example, the problematic path between the first two Union-Negates is as follows: We start from the Union-Negate in `l1`, then we have a Map, another Union (with no Negate), then we go to the first usage of `l1` (the second usage is fine, since it goes into a Join), which goes into the Union-Negate in `l3`.
+
+The Union-Negate pattern computes the outer part of a LEFT JOIN: the negated input to these unions is the collection of matched records. So, we have the most waste when all records have a match, in which case a consolidation would eliminate all records coming out from the Union-Negate. We can see that if we don't consolidate, then the record count increases by `2N`, where `N` is the size of one input, whereas if we consolidate, then record counts stay at `N`. Note that the increase is _not_ exponential (we have N, 3N, 5N, 7N, ...), but it will still get quite bad for big stacks. ([We've seen several big stacks](https://www.notion.so/materialize/Left-Join-Feedback-1d921cbe0296431b96b97b2480808cf9?d=e38c0c85241c48a98e8414f75e2c37e0). One customer [has a stack of 28 LEFT JOINs in one view, and 12 and 13 in two other views](https://materializeinc.slack.com/archives/C05GRCC4K1C/p1690293441722169).)
+
+Note that the problem occurs regardless of whether SemijoinIdempotence kicks in.
+
+# Explanation
+[explanation]: #explanation
+
+As mentioned above, our proposed solution is to simply perform a consolidation immediately after each of the Union-Negates. More specifically, during the MIR-to-LIR lowering of a Union, we'll check if the Union has an input that has a Negate at the top, and if yes, then we set a `consolidate` flag on the Union, which will make the rendering insert a `consolidate` call on the Union's result. This will immediately make the negated records cancel out.
+
+## Extra memory usage
+
+How `consolidate` works is that it buffers up data for each timestamp, and consolidates all data for the timestamp when the timestamp is closed. In steady state this won't result in much extra memory usage, but when processing a snapshot we will buffer up all the data temporarily. Note that if there are several LEFT JOINs, then only two of these `consolidate`s will simultaneously take up memory, because they are pipeline breakers.
```

Comment:
This is not entirely correct. The arrangement consolidates the input while it is reading data and sorts the data into power-of-two sized vectors, where each is compacted. It'll merge vectors to restore the power-of-two sized property while receiving data. This means that while receiving data, it'll already consolidate.

Reply:
Oh, ok, great, thanks! So then an extra consolidation introduces somewhat less extra memory use than I thought. I'll correct the text.

Reply:
I changed the text.